File: bms/go/ach/moov_io_ach_Addenda99Dishonored_TraceNumberField/moov_io_ach_Addenda99Dishonored_TraceNumberField.go
package test

import (
	"strings"
	"unicode/utf8"
)

var (
	moov_io_ach_stringZeros map[int]string = moov_io_ach_populateMap(94, "0")
)

const ()

func (Addenda99Dishonored *moov_io_ach_Addenda99Dishonored) TraceNumberField() string {
	return Addenda99Dishonored.stringField(Addenda99Dishonored.TraceNumber, 15)
}

func (c *moov_io_ach_converters) stringField(s string, max uint) string {
	ln := uint(utf8.RuneCountInString(s))
	if ln > max {
		return s[:max]
	}

	// Pad with preallocated string
	m := int(max - ln)
	pad, exists := moov_io_ach_stringZeros[m]
	if exists {
		return pad + s
	}
	// slow path
	return strings.Repeat("0", m) + s
}

func moov_io_ach_populateMap(max int, zero string) map[int]string {
	out := make(map[int]string, max)
	for i := 0; i < max; i++ {
		out[i] = strings.Repeat(zero, i)
	}
	return out
}

type moov_io_ach_Addenda99Dishonored struct {
	// ID is a client defined string used as a reference to this record.

	// TypeCode Addenda types code '99'

	// DishonoredReturnReasonCode is the return code explaining the dishonorment

	// OriginalEntryTraceNumber is the trace number specifieid in the initial entry

	// OriginalReceivingDFIIdentification is the DFI Identification specifieid in the initial entry

	// ReturnTraceNumber is the TraceNumber used when issuing the return

	// ReturnSettlementDate is the date of return issuing

	// ReturnReasonCode is the initial return code

	// AddendaInformation is additional data

	// TraceNumber is the trace number for dishonorment
	TraceNumber string `json:"traceNumber"`

	// validator is composed for data validation

	// converters is composed for ACH to GoLang Converters
	moov_io_ach_converters
}

type moov_io_ach_converters struct{}

File: translations_only/fail/go/ach/moov_io_ach_Addenda99Dishonored_TraceNumberField/mistral/Restart/moov_io_ach_Addenda99Dishonored_TraceNumberField.rs


use std::collections::HashMap;
use std::fmt;

struct MoovIoAchConverters {
    moov_io_ach_string_zeros: HashMap<usize, String>,
}

impl MoovIoAchConverters {
    fn string_field(&self, s: &str, max: usize) -> String {
        let ln = s.len();
        if ln > max {
            s[..max].to_string()
        } else {
            let m = max - ln;
            let pad = self
                .moov_io_ach_string_zeros
                .get(&m)
                .cloned()
                .unwrap_or_default();
            pad + s
        }
    }

    fn moov_io_ach_populate_map(max: usize, zero: char) -> HashMap<usize, String> {
        let mut out = HashMap::new();
        for i in 0..max {
            out.insert(i, "0".repeat(i));
        }
        out
    }
}

impl Default for MoovIoAchConverters {
    fn default() -> Self {
        MoovIoAchConverters {
            moov_io_ach_string_zeros: MoovIoAchConverters::moov_io_ach_populate_map(128, '0'),
        }
    }
}

#[derive(Default)]
struct MoovIoAchAddenda99Dishonored {
    trace_number: String,
    moov_io_ach_converters: Box<MoovIoAchConverters>,
}

impl MoovIoAchAddenda99Dishonored {
    fn trace_number_field(&self) -> String {
        self.string_field(&self.trace_number, 15)
    }

    fn string_field(&self, s: &str, max: usize) -> String {
        self.moov_io_ach_converters.string_field(s, max)
    }
}

impl fmt::Display for MoovIoAchAddenda99Dishonored {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "TraceNumber: {}",
            self.trace_number_field()
        )
    }
}

